### PR TITLE
week 7 hw

### PR DESCRIPTION
**Pull request by JolandaNava**

### CODE REVIEW

File: src/Week7/Editor.hs [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+modBuffer :: (b -> b) -> Editor b ()
+modBuffer = modify . first
+
+io :: MonadIO m => IO a -> m a
+io = liftIO
+
+-- Utility functions
+
+readMay :: Read a => String -> Maybe a
+readMay s = case reads s of
```
Comment (arbus) [inline src/Week7/Editor.hs:63]: There is already a `readMaybe` function in the module `Text.Read`

File: src/Week7/Scrabble.hs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+score 'x' = Score 8
+score 'y' = Score 4
+score 'z' = Score 10
+score _   = Score 0
+
+scoreString :: String -> Score
+scoreString = sum . map score
+
+getScore :: Score -> Int
+getScore (Score i) = i
```
Comment (arbus) [inline src/Week7/Scrabble.hs]: `Score` should probably be just a record with this name.

File: src/Week7/JoinList.hs [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+---------- ex 1 ----------
+tag :: Monoid m => JoinList m a -> m
+tag Empty          = mempty
+tag (Single m _)   = m
+tag (Append m _ _) = m
+
+(+++) :: Monoid m => JoinList m a -> JoinList m a -> JoinList m a
+(+++) Empty jl = jl
+(+++) jl Empty = jl
+(+++) jl1 jl2  = Append (tag jl1 <> tag jl2) jl1 jl2
```
Comment (chshersh) [inline src/Week7/JoinList.hs:25]: I would recommend implementing `Semigroup` and `Monoid` instances for `JoinList` to help you to understand concepts

File: src/Week7/Sized.hs [hunk 4]
```diff
@@ -0,0 +1,28 @@
+{-# LANGUAGE GeneralizedNewtypeDeriving, FlexibleInstances #-}
+module Week7.Sized where
+
+import Data.Monoid
+
+newtype Size = Size Int
+  deriving (Eq, Ord, Show, Num)
+
+getSize :: Size -> Int
+getSize (Size i) = i
```
Comment (chshersh) [inline src/Week7/Sized.hs]: Instead of writing this function separately you can just write `newtype Size = Size { getSize :: Int }`

**Review by chshersh — APPROVED** [review]
This looks great!

File: src/Week7/JoinList.hs [hunk 5]
```diff
@@ -24,6 +24,18 @@ tag (Append m _ _) = m
...
+instance Monoid m => Semigroup (JoinList m a) where
+    (<>) = (+++)
+-- does this qualify as associative?
+-- (a <> b) + c = a <> (b <> c)
+-- only if we consider this to be true
+-- Append m (Append n a b) c = Append m a (Append p b c)
+-- which they are if we understand them as list waiting to be appended
+
+instance Monoid m => Monoid (JoinList m a) where
+    mempty = Empty
```
Comment (JolandaNava) [inline src/Week7/JoinList.hs:37]: @ChShersh  I'm not sure this is what you had in mind

Comment (chshersh) [reply src/Week7/JoinList.hs:37]: @JolandaNava I had in mind exactly the same 👍 The problem with things like `Append m (Append n a b) c = Append m a (Append p b c)` is typical problem for such `Semigroup` instance. The general agreement that it's okay while your interface doesn't see difference between `Append m (Append n a b) c` and `Append m a (Append p b c)`

**Review by chshersh — APPROVED** [review]

**Pull request merged**